### DISCUSSION TITLE
Setting hours required

### DISCUSSION THREAD

**Discussion by djdarlo** [BottlecapDave/HomeAssistant-OctopusEnergy#206] [Q&A]
I'm going to use this cool integration to charge my batteries in the early hours of the morning when the energy is cheaper. 

The one thing I want to do is evaluate at say 11:30pm the night before roughly how much charge I'll need to do. Something like average daily load (15kWh) - current battery SoC - solar forecast for the next day. 

Based on the result of the above I'll work out how many half hour slots I need. Roughly 30mins per 2.5kWh. So if the result is 5kWh I'll need 1 hour. 7.5kWh, 1.5hours etc etc. 

So I plan to set a target rate period from 00:00 to 06:00, intermittent the only variable I want to change is the number of hours depending on my calculation. 

Is this easy to set?

**Comment by BottlecapDave** [comment]
Hello and welcome!

You can certainly do what you're requesting by using the available [update service](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy#service-octopus_energyupdate_target_config). You can set just the number of target hours which will then update the sensor and re-evaluate it. Just like when you set up the target rate sensor, hours need to be in half hour increments.

**Reply by djdarlo** [reply]
I've tried and think my syntax is letting me down.  Any ideas?

alias: Octopus Battery Boost
description: >-
  Looks at current SoC, tomorrow's Solar Forecast and works out how long to
  charge the battery for.  30mins = 2.5kWh.
trigger:
  - platform: time
    at: "14:05:00"
condition: []
action:
  - service: octopus_energy.update_target_config
    data: >
      {% if (15 - (states('sensor.solcast_forecast_tomorrow') |
      float(default=0)) - (states('sensor.bms_kwh_remaining_2') |
      float(default=0))) <= 2.5 %}
        {"target_hours": "10.0", "target_start_time": "11:00", "target_end_time": "21:00"}
      {% endif %}

      {% if (15 - (states('sensor.solcast_forecast_tomorrow') |
      float(default=0)) - (states('sensor.bms_kwh_remaining_2') |
      float(default=0))) > 2.5 %}
        {"target_hours": "0.5" }
      {% endif %}

      {% if (15 - (states('sensor.solcast_forecast_tomorrow') |
      float(default=0)) - (states('sensor.bms_kwh_remaining_2') |
      float(default=0))) > 5 %}
        {"target_hours": "1.0" }
      {% endif %}

      {% if (15 - (states('sensor.solcast_forecast_tomorrow') |
      float(default=0)) - (states('sensor.bms_kwh_remaining_2') |
      float(default=0))) > 7.5 %}
        {"target_hours": "1.5" }
      {% endif %}

      {% if (15 - (states('sensor.solcast_forecast_tomorrow') |
      float(default=0)) - (states('sensor.bms_kwh_remaining_2') |
      float(default=0))) > 10 %}
        {"target_hours": "2.0" }
      {% endif %} 
    target:
      entity_id: binary_sensor.octopus_energy_target_octopus_half_hourly_target_rates
mode: single

I get the following error

Triggered manually at April 9, 2023 at 2:12:30 PM
Octopus Energy: Update target rate config Octopus Energy Target octopus_half_hourly_target_rates
Stopped because an error was encountered at April 9, 2023 at 2:12:30 PM (runtime: 0.04 seconds)

division by zero

**Reply by BottlecapDave** [reply]
No problem. You can adapt the example with the service in the [readme](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy#service-octopus_energyupdate_target_config) to look like the above (I believe)

```
trigger:
  - platform: time
    at: "14:05:00"
  condition: []
  action:
  - service: octopus_energy.update_target_config
    data:
      target_hours: >
      {% if (15 - (states('sensor.solcast_forecast_tomorrow') | float(default=0)) - (states('sensor.bms_kwh_remaining_2') | float(default=0))) > 10 %}
      2.0
      {% elif (15 - (states('sensor.solcast_forecast_tomorrow') | float(default=0)) - (states('sensor.bms_kwh_remaining_2') | float(default=0))) > 7.5 %}
      1.5
      {% elif (15 - (states('sensor.solcast_forecast_tomorrow') | float(default=0)) - (states('sensor.bms_kwh_remaining_2') | float(default=0))) > 5 %}
      1.0
      {% elif (15 - (states('sensor.solcast_forecast_tomorrow') | float(default=0)) - (states('sensor.bms_kwh_remaining_2') | float(default=0))) > 2.5 %}
      0.5
      {% elif (15 - (states('sensor.solcast_forecast_tomorrow') | float(default=0)) - (states('sensor.bms_kwh_remaining_2') | float(default=0))) <= 2.5 %}
      10.0
      {% endif %}
  target:
    entity_id: binary_sensor.octopus_energy_target_octopus_half_hourly_target_rates
```

because of how you're looking at remaining percentage, you need to reverse the checks as multiple statements can be true and only the first if statement will be picked (e.g. if the value is 6, then the first three if statements will be true but only the first if statement will be picked resulting in a value of 0.5

**Comment by djdarlo** [comment]
Thank-you.  I got it working last night and it's working a treat.  This morning I was trying to use some of the variables in the target config set to do some other automation but I am getting the following error message.  The target config set is between midnight and 6am therefore will this correct itself when tomorrow's agile rates are published around 4pm?

This entity is no longer being provided by the octopus_energy integration. If the entity is no longer in use, delete it in settings.

**Reply by BottlecapDave** [reply]
As per your issue, this should be fixed in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/v6.6.1
